### ISSUE TITLE
[core][credentials] Adding credentials should handle wrong ids

### ISSUE THREAD

**Issue by egekorkan** [eclipse/thingweb.node-wot#277]
Currently, the line that adds the credential for each Thing based on their `id` does not handle wrong `id`s given by the developer and when the credential for the Thing based on the `id` of the TD is gotten, it crashes with generic error message that doesn't help to spot the error (cannot read property of undefined). There should be a try/catch or another construct around it that gives an error message like "id provided in the credentials does not correspond to the `id` of the Thing in its TD". I think that this error handling needs to happen when they are gotten since when adding the credentials, there may be no TD produced.

Where the credentials are stored:
https://github.com/eclipse/thingweb.node-wot/blob/d647247fe14d2384653910699c3324597f606282/packages/core/src/servient.ts#L276

Where they are gotten:
https://github.com/eclipse/thingweb.node-wot/blob/d647247fe14d2384653910699c3324597f606282/packages/core/src/servient.ts#L282

**Comment by danielpeintner** [comment]
@egekorkan I agree, the line

`return this.credentialStore.get(identifier); `

could be simply replaced by something like

        if (this.credentialStore.has(identifier)) {
            return this.credentialStore.get(identifier);
        } else {
            // what whould the expectation?
            return null;
        }

I just wonder *what* would be the expectation?
* returning null
* throwing error
* ?

BTW, while looking at the code I found a different issue, see https://github.com/eclipse/thingweb.node-wot/blob/d647247fe14d2384653910699c3324597f606282/packages/core/src/servient.ts#L220-L224

it uses variable `name` while it should use `id` instead. Am I wrong?

**Comment by egekorkan** [comment]
I think throwing `null` makes sense :)

**Comment by danielpeintner** [comment]
see PR https://github.com/eclipse/thingweb.node-wot/pull/283, please approve @egekorkan

**Comment by danielpeintner** [comment]
PR https://github.com/eclipse/thingweb.node-wot/pull/283 has been closed as discussed.
I suggest we keep the issue open until we have checked the behavior of each binding.

**Comment by egekorkan** [comment]
Another related point to this: 
It is also not possible to assign different security schemes to a single id. For example, if I want to give access to two Consumers with basic auth but not give the same username password to them, it is not possible to do it :/

**Comment by danielpeintner** [comment]
> It is also not possible to assign different security schemes to a single id. For example, if I want to give access to two Consumers with basic auth but not give the same username password to them, it is not possible to do it :/

This is correct.
However such a change would be a *breaking change* at least for `getCredentials()` unless we deprecate _this_ method and create a new one getting multiple credentials. Also, when `addCredentials(credentials: any)` is possible several times we need also a call for removing credentials.
Nor sure if such a change is worth it unless we have an actual need...

**Comment by egekorkan** [comment]
I think that `addCredentials` should firstly add new credentials, currently it is setting the credentials. I would also support a method like `removeCredentials`. Also the word `getCredentials` implies that there can be multiple credentials returned, which is not possible at the moment.

Use case: We proxy node-wot implemented devices that are in a local network and cannot use node-wot as the proxy since the consumers need individual credentials per device.

**Comment by danielpeintner** [comment]
I agree with your statements but I do not think we can for example *fix*  `getCredentials` by returning multiple credentials. Hence I suggest to mark it as `@deprecated` (and keep the current return value) and create a new method like `getAllCredentials` which may return multiple credentials.

Does this work for @egekorkan, others? If so I can create a PR accordingly...

**Comment by relu91** [comment]
I agree, then when we publish a major version of node-wot I'd remove all the deprecated methods. However, `getAllCredentials` sounds a little bit weird to me. What about to just change the name to `retrieveCredentials` ?

**Comment by danielpeintner** [comment]
> What about to just change the name to `retrieveCredentials` ?

That sounds better, indeed ;-)

**Comment by egekorkan** [comment]
I also prefer `retrieveCredentials` :+1:

**Comment by danielpeintner** [comment]
I started a PR https://github.com/eclipse/thingweb.node-wot/pull/300

Some issues I see or I run into
* how to properly remove a dedicated credential (we would need to do a diff to find the *right* one). Maybe just support `resetCredentials()` ???
* `interface ProtocolClient` in `protocol-interfaces.ts` has `setSecurity(metadata: Array<TD.SecurityScheme>, credentials?: any): boolean`
  We would need to change all these methods also, right (passing array of credentials?)? Or *who* decides which credential to pick?

**Comment by danielpeintner** [comment]
@relu91 do you have any comments w.r.t. the [issues above](https://github.com/eclipse/thingweb.node-wot/issues/277#issuecomment-668481823)

**Comment by relu91** [comment]
> how to properly remove a dedicated credential (we would need to do a diff to find the right one). Maybe just support `resetCredentials()` ???

I agree that removing a single credential is tricky. Maybe we can think a little bit better about a credential data model and using a proper type instead of `any`.  So that we can have an equality method to check if two credentials are the same.  In the meantime defining `resetCredentials` is the best option; users of the library can combine `resetCredentials' with `retriveCredentials` and `addCredentials`. It is more verbose but we cover the "remove a credential" use case. 

> We would need to change all these methods also, right (passing array of credentials?)? Or who decides which credential to pick?

yes. however, I do not know how to decide which credential to pick from a protocol binding implementation. I'm afraid we should define another field in [interaction options](https://w3c.github.io/wot-scripting-api/#dom-interactionoptions); like a `credentialIndex` (?).

**Comment by egekorkan** [comment]
> yes. however, I do not know how to decide which credential to pick from a protocol binding implementation. I'm afraid we should define another field in interaction options; like a credentialIndex (?).

Even though this seems the only option, it indeed makes things more complicated and requires even more fine grained TD knowledge by the developer

**Comment by danielpeintner** [comment]
> > yes. however, I do not know how to decide which credential to pick from a protocol binding implementation. I'm afraid we should define another field in interaction options; like a credentialIndex (?).
> 
> Even though this seems the only option, it indeed makes things more complicated and requires even more fine grained TD knowledge by the developer

The best approach (it seems) is like we do for `formIndex`. IF there is any hint use it. If there is no hint just pick the first one ....

**Comment by relu91** [comment]
> The best approach (it seems) is like we do for formIndex. IF there is any hint use it. If there is no hint just pick the first one ....

The problem that I see here, it is that credentials are no a TD concept (on the contrary of forms). So it might be confusing specify such field in the API spec. We have first to tackle this problem consistently, I fear that leaving credential management completely to runtime it might be too complex...

**Comment by danielpeintner** [comment]
Note: we may want to add decorators like `@deprecated` to indicate obsolete functions
(see also https://medium.com/@rossbulat/get-started-with-typescript-decorators-cf3924c37f04)
